src/main/java/org/folio/circulation/support/http/client/ResponseInterpreter.java

```diff
@@ -0,0 +1,79 @@
+package org.folio.circulation.support.http.client;
+
+import static org.folio.circulation.support.Result.failed;
+
+import java.lang.invoke.MethodHandles;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.folio.circulation.support.Result;
+import org.folio.circulation.support.ServerErrorFailure;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ResponseInterpreter<T> {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private final Map<Integer, Function<Response, Result<T>>> responseMappers;
+  private final Function<Response, Result<T>> unexpectedResponseMapper;
+
+  private ResponseInterpreter(Map<Integer, Function<Response, Result<T>>> responseMappers,
+                              Function<Response, Result<T>> unexpectedResponseMapper) {
+
+    this.unexpectedResponseMapper = unexpectedResponseMapper;
+    this.responseMappers = responseMappers;
+  }
+
+  public ResponseInterpreter() {
+    this(new HashMap<>(), ResponseInterpreter::defaultUnexpectedResponseMapper);
+  }
+
+  public ResponseInterpreter<T> flatMapOn(Integer status,
+                                          Function<Response, Result<T>> mapper) {
+
+    final HashMap<Integer, Function<Response, Result<T>>> newMappers
+      = new HashMap<>(responseMappers);
+
+    newMappers.put(status, mapper);
+
+    return new ResponseInterpreter<>(newMappers, unexpectedResponseMapper);
+  }
+
+  public ResponseInterpreter<T> on(Integer status, Result<T> result) {
+    return flatMapOn(status, response -> result);
+  }
+
+  public ResponseInterpreter<T> otherwise(
+    Function<Response, Result<T>> unexpectedResponseMapper) {
+
+    return new ResponseInterpreter<>(responseMappers, unexpectedResponseMapper);
+  }
+
+  public Result<T> apply(Response response) {
+    if (response == null) {
+      log.warn("Cannot interpret null response");
+      return failed(new ServerErrorFailure("Cannot interpret null response"));
+    }
+
+    try {
+      log.info("Response received: {}", response);
+
+      final Integer statusCode = response.getStatusCode();
+
+      return responseMappers.getOrDefault(statusCode, unexpectedResponseMapper)
+        .apply(response);
+    }
+    catch (Exception e) {
+      return failed(e);
```

Comment:
~~Should this exception be wrapped in a `ServerErrorFailure`?~~

Oops, I decided, after reviewing the unit tests where the exceptions appear to be wrapped in `ServerErrorFailure`, to see what `failed` does and I see that `Result.falied(Throwable)` does indeed wrap the exception in a `ServerErrorFailure`.

Since this lead to some confusion, at least for me, when looking at this code in isolation, perhaps we could add a `Result.failed(String)` that wraps the `String` message in a `ServerErrorFailure`? That way, the `failed` calls are a bit more consistent in this class. No big deal either way, just took a few minutes extra to sort this out from my point of view.

Reply:
I agree this is confusing. I think in part, this is because the Result class is over-reaching a little to try to be helpful.

I think we can add another overload to `failed`. 

Alternatively, we could introduce a more explicit method for this, e.g. `failedDueToServerError` and also migrate the exception variant to that as well. It would reduce the size of the `Result` itself as well.

What do you think?

Reply:
Part of this is also a deficiency in the design of the result and failure classes (they try too hard to make output for them easy). 

Partly due to Java 8's lack of choice types (or a better type based switch, which I think has been proposed).

Reply:
This is really good feedback, it's great to know what areas are confusing.

Reply:
A `failedDueToServerError` makes sense to me.

Reply:
I've added a couple of support methods (into a new results namespace, where I'd like to move some other classes to at some point).

Is it less confusing?